hooks/command

```diff
@@ -221,34 +172,49 @@ echo "--- Getting the password to unlock the keychain"
 keychain_pw_helper_script=$(find_keychain_pw_helper_script)
 keychain_pw=$(get_keychain_pw "${keychain_pw_helper_script}")
 
-echo "--- Unlocking the keychain"
+#echo "--- Unlocking the keychain"
 unlock_keychain "${codesigning_keychain}" "${keychain_pw}"
 
 echo "--- Setting up local directories"
 # Download things in a local dir to identify them
 artifact_dir_fragment="artifacts"
 rm -rf "${artifact_dir_fragment}" && mkdir -p "${artifact_dir_fragment}"
-relative_artifacts_dir="$(pwd)/${artifact_dir_fragment}"
-
-# Signed things go in a local dir so the uploaded artifacts don't have a weird path
-signed_dir_fragment="signed"
-rm -rf "${signed_dir_fragment}" && mkdir -p "${signed_dir_fragment}"
-
-# Download and sign each requested artifact
-for artifact in $(plugin_read_list INPUT_ARTIFACTS); do
-  echo "--- Signing ${artifact}"
+absolute_artifacts_dir="$(pwd)/${artifact_dir_fragment}"
 
+## Download and sign each requested artifact
+plugin_read_list INPUT_ARTIFACTS | while read -r artifact; do
   echo "${artifact}: retrieving unsigned artifact"
-  unsigned_artifact="$(fetch_artifact "${artifact}" "${relative_artifacts_dir}")"
-
-  sign_and_notarize "${unsigned_artifact}" "${signed_dir_fragment}"
-  retval=$?
-  if [[ "${retval}" -ne 0 ]]; then
-    echo "failed to sign and notarize '${unsigned_artifact}': error code '${retval}'"
-    exit ${retval}
-  fi
-
+  fetch_artifact "${artifact}" "${absolute_artifacts_dir}"
 done
 
-echo "--- Uploading artifacts"
-upload_artifact "${signed_dir_fragment}/**/*"
+entitlements="${BUILDKITE_PLUGIN_MAC_CODESIGN_ENTITLEMENTS}"
+if [[ -n "${entitlements}" ]]; then
+  #  fetch_artifact "${entitlements}" "${absolute_artifacts_dir}"
+  entitlements="${absolute_artifacts_dir}/${entitlements}"
+fi
+
+config_path="config.json"
+prerequisites="$(plugin_read_list SIGN_PREREQUISITES)"
+
+# generate config
+echo "-- Generating signing config"
+generate_signing_config "${prerequisites}" "${entitlements}" "${absolute_artifacts_dir}" "${config_path}"
+
+echo "-- Executing gon on generated config"
+gon "${config_path}"
+retval=$?
+if [[ "${retval}" -ne 0 ]]; then
+  echo "notarization failed: error code '${retval}'"
+  exit 5
```

Comment:
Would it make sense to maybe return a shifted error code here?

EG, instead of always returning `5` on notarization errors, maybe we do something like

```suggestion
  NOTARIZATION_ERR_OFFSET=16
  exit ${NOTARIZATION_ERR_OFFSET} + ${retval}
```

That way individual notarization errors would be individually distinguishable in logs (eg notarization returns 1 => `exit 16`, notarization returns 2 => `exit 17`, etc).

Totally not a requirement here, and I don't even know if we get back varying error codes from notarization errors anyway (👊 🍎 ), so this might all just be a moot point anyway.

Reply:
Id be fine with adding this, but i just realised our use of 
```
set -o errexit -o nounset -o pipefail
[[ -n "${DEBUG-}" ]] && set -x
```
means that these if statements never actually get triggered. Woops.

Reply:
Whoops indeed. That's what I get for copy/pasting setup from elsewhere and not taking `pipefail` into account. 🤦